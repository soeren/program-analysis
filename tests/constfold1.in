// Test for constant folding:
// - Check if arithmetical expressions containing constants are folded
// - Check that boolean expressions containing constants in IF or
// WHILE are folded
// - Check correctness of constant propagation in/after WHILE and IF statements

program

int x;
int y;
int z;

x := 23;
y := 42;
z := x*2; // should be 46

if(z > y){ // should be folded to if(true)
    y := 0;
} else {
    y := 1;
}

if(y = 0){ // should stay the same (y not a const anymore)
    y := 1;
} else {
    y := 0;
}

while(y > 0){
    while(z < 0){ // z stays const during while -> folding to false
        y := y +1;
    }
    y := y -1;
}

while(z > 0){ // z not a const in WHILE loop -> stay the same
    z := z -1;
    if(z = x){ // x is const -> folding to if(z=23)
        y := 0;
    } else {
        y := 1;
    }
}

while(z > 0){ // no more consts, stays the same
    z := z -1;
    if(z = x){
        x := 23;
    } else {
        x := 42;
    }
}


write(x);

end
