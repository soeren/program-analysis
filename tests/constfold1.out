program
int x;
int y;
int z;
x := 23; // 1
y := 42; // 2
z := 46; // 3
if(true) // 4
{
     y := 0; // 5
 } else
{
     y := 1; // 6
 }
if(y = 0) // 7
{
     y := 1; // 8
 } else
{
     y := 0; // 9
 }
while(y > 0) // 10
{
     while(false) // 11
     {
          y := y + 1; // 12
      }
     y := y - 1; // 13
 }
while(z > 0) // 14
{
     z := z - 1; // 15
     if(z = 23) // 16
     {
          y := 0; // 17
      } else
     {
          y := 1; // 18
      }
 }
while(z > 0) // 19
{
     z := z - 1; // 20
     if(z = x) // 21
     {
          x := 23; // 22
      } else
     {
          x := 42; // 23
      }
 }
write(x); // 24
end
